File: app/src/main/java/com/adonai/manman/adapters/LocalArchiveArrayAdapter.java
package com.adonai.manman.adapters;

import android.content.Context;
import android.text.TextUtils;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.Filter;
import android.widget.ImageView;
import android.widget.TextView;

import com.adonai.manman.R;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

/**
 * Array adapter for showing files in local man page archive
 * The data retrieval is done through {@link com.adonai.manman.ManLocalArchiveFragment.LocalArchiveParserCallback}
 *
 * @see android.widget.ArrayAdapter
 * @see java.io.File
 * @author Adonai
 */
public class LocalArchiveArrayAdapter extends ArrayAdapter<File> {
    private List<File> originals;
    private List<File> filtered;

    public LocalArchiveArrayAdapter(Context context, int resource, int textViewResourceId, List<File> objects) {
        super(context, resource, textViewResourceId, objects);
        originals = objects;
        filtered = objects;
    }

    @Override
    public int getCount() {
        return filtered.size();
    }

    @Override
    public File getItem(int position) {
        return filtered.get(position);
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        final File current = getItem(position);
        String effectiveName = current.getName().replaceAll("\\.gz$", "").replaceAll("\\.\\d$", "");
        View root = super.getView(position, convertView, parent);

        TextView command = (TextView) root.findViewById(R.id.command_name_label);
        command.setText(effectiveName);

        TextView url = (TextView) root.findViewById(R.id.command_description_label);
        url.setText(current.getParent());

        ImageView popup = (ImageView) root.findViewById(R.id.popup_menu);
        popup.setVisibility(View.GONE); // save for future, hide for now

        return root;
    }

    @Override
    public Filter getFilter() {
        return new Filter() {

            @Override
            protected FilterResults performFiltering(CharSequence constraint) {
                FilterResults fr = new FilterResults();
                if(TextUtils.isEmpty(constraint)) { // special case for empty filter
                    fr.values = originals;
                    fr.count = originals.size();
                    return fr;
                }

                List<File> tempFilteredValues = new ArrayList<>();
                for(File f : originals) {
                    if(f.getName().startsWith(constraint.toString())) {
                        tempFilteredValues.add(f);
                    }
                }

                fr.values = tempFilteredValues;
                fr.count = tempFilteredValues.size();
                return fr;
            }

            @SuppressWarnings("unchecked")
            @Override
            protected void publishResults(CharSequence constraint, FilterResults results) {
                filtered = (List<File>) results.values;
                notifyDataSetChanged();
            }
        };
    }
}

File: app/src/main/java/com/adonai/manman/adapters/LocalArchiveArrayAdapter.kt
package com.adonai.manman.adapters

import android.content.Context
import android.text.TextUtils
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.Filter
import android.widget.ImageView
import android.widget.TextView
import com.adonai.manman.R
import com.adonai.manman.ManLocalArchiveFragment
import java.io.File
import java.util.*

/**
 * Array adapter for showing files in local man page archive
 * The data retrieval is done through [ManLocalArchiveFragment.doLoadContent]
 *
 * @see ArrayAdapter
 * @see File
 *
 * @author Kanedias
 */
class LocalArchiveArrayAdapter(context: Context, resource: Int, textViewResourceId: Int, private val originals: List<File>) : ArrayAdapter<File>(context, resource, textViewResourceId, originals) {

    private var filtered: List<File> = originals

    override fun getCount(): Int {
        return filtered.size
    }

    override fun getItem(position: Int): File {
        return filtered[position]
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        val root = super.getView(position, convertView, parent)
        val current = getItem(position)
        val effectiveName = current.name.replace("\\.gz$".toRegex(), "").replace("\\.\\d$".toRegex(), "")

        val command = root.findViewById<View>(R.id.command_name_label) as TextView
        val url = root.findViewById<View>(R.id.command_description_label) as TextView
        val popup = root.findViewById<View>(R.id.popup_menu) as ImageView

        command.text = effectiveName
        url.text = current.parent
        popup.visibility = View.GONE // save for future, hide for now
        return root
    }

    override fun getFilter(): Filter {
        return object : Filter() {
            override fun performFiltering(constraint: CharSequence): FilterResults {
                val fr = FilterResults()

                if (TextUtils.isEmpty(constraint)) { // special case for empty filter
                    fr.values = originals
                    fr.count = originals.size
                    return fr
                }

                val tempFilteredValues: MutableList<File> = ArrayList()
                for (f in originals) {
                    if (f.name.startsWith(constraint.toString())) {
                        tempFilteredValues.add(f)
                    }
                }

                fr.values = tempFilteredValues
                fr.count = tempFilteredValues.size
                return fr
            }

            override fun publishResults(constraint: CharSequence, results: FilterResults) {
                filtered = results.values as List<File>
                notifyDataSetChanged()
            }
        }
    }
}